File: results/0.25/ZIP/zip-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

#define LOCAL_FILE_HEADER_SIGNATURE 0x04034b50
#define CENTRAL_DIRECTORY_HEADER_SIGNATURE 0x02014b50
#define END_OF_CENTRAL_DIRECTORY_SIGNATURE 0x06054b50

typedef struct {
    uint32_t signature;
    uint16_t version_needed_to_extract;
    uint16_t general_purpose_bit_flag;
    uint16_t compression_method;
    uint16_t last_modified_time;
    uint16_t last_modified_date;
    uint32_t crc_32;
    uint32_t compressed_size;
    uint32_t uncompressed_size;
    uint16_t file_name_length;
    uint16_t extra_field_length;
} local_file_header_t;

typedef struct {
    uint32_t signature;
    uint16_t version_made_by;
    uint16_t version_needed_to_extract;
    uint16_t general_purpose_bit_flag;
    uint16_t compression_method;
    uint16_t last_modified_time;
    uint16_t last_modified_date;
    uint32_t crc_32;
    uint32_t compressed_size;
    uint32_t uncompressed_size;
    uint16_t file_name_length;
    uint16_t extra_field_length;
    uint16_t file_comment_length;
    uint16_t disk_number_start;
    uint16_t internal_attributes;
    uint32_t external_attributes;
} central_directory_header_t;

typedef struct {
    uint32_t signature;
    uint16_t number_of_this_disk;
    uint16_t number_of_the_disk_where_the_central_directory_starts;
    uint16_t number_of_entries_in_the_central_directory_on_this_disk;
    uint16_t total_number_of_entries_in_the_central_directory;
    uint32_t size_of_the_central_directory;
    uint32_t offset_of_the_start_of_the_central_directory;
    uint16_t comment_length;
} end_of_central_directory_t;

typedef struct {
    uint16_t header_id;
    uint16_t data_size;
    uint8_t* data;
} extra_field_t;

typedef struct {
    local_file_header_t* local_file_header;
    uint8_t* file_name;
    extra_field_t* extra_field;
    uint8_t* file_data;
} local_file_t;

typedef struct {
    central_directory_header_t* central_directory_header;
    uint8_t* file_name;
    extra_field_t* extra_field;
    uint8_t* file_comment;
} central_directory_t;

typedef struct {
    end_of_central_directory_t* end_of_central_directory;
    uint8_t* comment;
} end_of_central_directory_record_t;

void local_file_header_parser(void* parser, local_file_header_t* local_file_header) {
    H_SEQUENCE(
        H_BIND_UINT32_LE(H_CHECK(LOCAL_FILE_HEADER_SIGNATURE), local_file_header->signature),
        H_BIND_UINT16_LE(H_CHECK(0x10), local_file_header->version_needed_to_extract),
        H_BIND_UINT16_LE(H_CHECK(0x0000), local_file_header->general_purpose_bit_flag),
        H_BIND_UINT16_LE(H_CHECK(0x0000), local_file_header->compression_method),
        H_BIND_UINT16_LE(H_CHECK(0x0000), local_file_header->last_modified_time),
        H_BIND_UINT16_LE(H_CHECK(0x0000), local_file_header->last_modified_date),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), local_file_header->crc_32),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), local_file_header->compressed_size),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), local_file_header->uncompressed_size),
        H_BIND_UINT16_LE(H_CHECK(0x0000), local_file_header->file_name_length),
        H_BIND_UINT16_LE(H_CHECK(0x0000), local_file_header->extra_field_length)
    );
}

void extra_field_parser(void* parser, extra_field_t* extra_field) {
    H_SEQUENCE(
        H_BIND_UINT16_LE(H_CHECK(0x0001), extra_field->header_id),
        H_BIND_UINT16_LE(H_CHECK(0x0000), extra_field->data_size),
        H_BIND_BYTES(extra_field->data_size, extra_field->data)
    );
}

void local_file_parser(void* parser, local_file_t* local_file) {
    H_SEQUENCE(
        H_BIND_STRUCT(local_file_header_parser, local_file->local_file_header),
        H_BIND_BYTES(local_file->local_file_header->file_name_length, local_file->file_name),
        H_BIND_STRUCT(extra_field_parser, local_file->extra_field),
        H_BIND_BYTES(local_file->local_file_header->compressed_size, local_file->file_data)
    );
}

void central_directory_header_parser(void* parser, central_directory_header_t* central_directory_header) {
    H_SEQUENCE(
        H_BIND_UINT32_LE(H_CHECK(CENTRAL_DIRECTORY_HEADER_SIGNATURE), central_directory_header->signature),
        H_BIND_UINT16_LE(H_CHECK(0x10), central_directory_header->version_made_by),
        H_BIND_UINT16_LE(H_CHECK(0x10), central_directory_header->version_needed_to_extract),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->general_purpose_bit_flag),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->compression_method),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->last_modified_time),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->last_modified_date),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), central_directory_header->crc_32),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), central_directory_header->compressed_size),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), central_directory_header->uncompressed_size),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->file_name_length),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->extra_field_length),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->file_comment_length),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->disk_number_start),
        H_BIND_UINT16_LE(H_CHECK(0x0000), central_directory_header->internal_attributes),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), central_directory_header->external_attributes)
    );
}

void central_directory_parser(void* parser, central_directory_t* central_directory) {
    H_SEQUENCE(
        H_BIND_STRUCT(central_directory_header_parser, central_directory->central_directory_header),
        H_BIND_BYTES(central_directory->central_directory_header->file_name_length, central_directory->file_name),
        H_BIND_STRUCT(extra_field_parser, central_directory->extra_field),
        H_BIND_BYTES(central_directory->central_directory_header->file_comment_length, central_directory->file_comment)
    );
}

void end_of_central_directory_parser(void* parser, end_of_central_directory_record_t* end_of_central_directory_record) {
    H_SEQUENCE(
        H_BIND_UINT32_LE(H_CHECK(END_OF_CENTRAL_DIRECTORY_SIGNATURE), end_of_central_directory_record->end_of_central_directory->signature),
        H_BIND_UINT16_LE(H_CHECK(0x0000), end_of_central_directory_record->end_of_central_directory->number_of_this_disk),
        H_BIND_UINT16_LE(H_CHECK(0x0000), end_of_central_directory_record->end_of_central_directory->number_of_the_disk_where_the_central_directory_starts),
        H_BIND_UINT16_LE(H_CHECK(0x0000), end_of_central_directory_record->end_of_central_directory->number_of_entries_in_the_central_directory_on_this_disk),
        H_BIND_UINT16_LE(H_CHECK(0x0000), end_of_central_directory_record->end_of_central_directory->total_number_of_entries_in_the_central_directory),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), end_of_central_directory_record->end_of_central_directory->size_of_the_central_directory),
        H_BIND_UINT32_LE(H_CHECK(0x00000000), end_of_central_directory_record->end_of_central_directory->offset_of_the_start_of_the_central_directory),
        H_BIND_UINT16_LE(H_CHECK(0x0000), end_of_central_directory_record->end_of_central_directory->comment_length),
        H_BIND_BYTES(end_of_central_directory_record->end_of_central_directory->comment_length, end_of_central_directory_record->comment)
    );
}

int main(int argc, char** argv) {
    if (argc != 2) {
        printf("Usage: %s <zip_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        printf("Error opening file %s\n", argv[1]);
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* data = malloc(file_size);
    if (!data) {
        printf("Error allocating memory\n");
        fclose(file);
        return 1;
    }

    size_t bytes_read = fread(data, 1, file_size, file);
    if (bytes_read != file_size) {
        printf("Error reading file\n");
        free(data);
        fclose(file);
        return 1;
    }

    fclose(file);

    local_file_t local_file;
    central_directory_t central_directory;
    end_of_central_directory_record_t end_of_central_directory_record;

    int result = H_PARSE(local_file_parser, data, file_size, &local_file);
    if (result == 0) {
        printf("Local file header:\n");
        printf("  Signature: 0x%08x\n", local_file.local_file_header->signature);
        printf("  Version needed to extract: 0x%04x\n", local_file.local_file_header->version_needed_to_extract);
        printf("  General purpose bit flag: 0x%04x\n", local_file.local_file_header->general_purpose_bit_flag);
        printf("  Compression method: 0x%04x\n", local_file.local_file_header->compression_method);
        printf("  Last modified time: 0x%04x\n", local_file.local_file_header->last_modified_time);
        printf("  Last modified date: 0x%04x\n", local_file.local_file_header->last_modified_date);
        printf("  CRC-32: 0x%08x\n", local_file.local_file_header->crc_32);
        printf("  Compressed size: 0x%08x\n", local_file.local_file_header->compressed_size);
        printf("  Uncompressed size: 0x%08x\n", local_file.local_file_header->uncompressed_size);
        printf("  File name length: 0x%04x\n", local_file.local_file_header->file_name_length);
        printf("  Extra field length: 0x%04x\n", local_file.local_file_header->extra_field_length);
    } else {
        printf("Error parsing local file header\n");
    }

    result = H_PARSE(central_directory_parser, data, file_size, &central_directory);
    if (result == 0) {
        printf("Central directory header:\n");
        printf("  Signature: 0x%08x\n", central_directory.central_directory_header->signature);
        printf("  Version made by: 0x%04x\n", central_directory.central_directory_header->version_made_by);
        printf("  Version needed to extract: 0x%04x\n", central_directory.central_directory_header->version_needed_to_extract);
        printf("  General purpose bit flag: 0x%04x\n", central_directory.central_directory_header->general_purpose_bit_flag);
        printf("  Compression method: 0x%04x\n", central_directory.central_directory_header->compression_method);
        printf("  Last modified time: 0x%04x\n", central_directory.central_directory_header->last_modified_time);
        printf("  Last modified date: 0x%04x\n", central_directory.central_directory_header->last_modified_date);
        printf("  CRC-32: 0x%08x\n", central_directory.central_directory_header->crc_32);
        printf("  Compressed size: 0x%08x\n", central_directory.central_directory_header->compressed_size);
        printf("  Uncompressed size: 0x%08x\n", central_directory.central_directory_header->uncompressed_size);
        printf("  File name length: 0x%04x\n", central_directory.central_directory_header->file_name_length);
        printf("  Extra field length: 0x%04x\n", central_directory.central_directory_header->extra_field_length);
        printf("  File comment length: 0x%04x\n", central_directory.central_directory_header->file_comment_length);
        printf("  Disk number start: 0x%04x\n", central_directory.central_directory_header->disk_number_start);
        printf("  Internal attributes: 0x%04x\n", central_directory.central_directory_header->internal_attributes);
        printf("  External attributes: 0x%08x\n", central_directory.central_directory_header->external_attributes);
    } else {
        printf("Error parsing central directory header\n");
    }

    result = H_PARSE(end_of_central_directory_parser, data, file_size, &end_of_central_directory_record);
    if (result == 0) {
        printf("End of central directory record:\n");
        printf("  Signature: 0x%08x\n", end_of_central_directory_record.end_of_central_directory->signature);
        printf("  Number of this disk: 0x%04x\n", end_of_central_directory_record.end_of_central_directory->number_of_this_disk);
        printf("  Number of the disk where the central directory starts: 0x%04x\n", end_of_central_directory_record.end_of_central_directory->number_of_the_disk_where_the_central_directory_starts);
        printf("  Number of entries in the central directory on this disk: 0x%04x\n", end_of_central_directory_record.end_of_central_directory->number_of_entries_in_the_central_directory_on_this_disk);
        printf("  Total number of entries in the central directory: 0x%04x\n", end_of_central_directory_record.end_of_central_directory->total_number_of_entries_in_the_central_directory);
        printf("  Size of the central directory: 0x%08x\n", end_of_central_directory_record.end_of_central_directory->size_of_the_central_directory);
        printf("  Offset of the start of the central directory: 0x%08x\n", end_of_central_directory_record.end_of_central_directory->offset_of_the_start_of_the_central_directory);
        printf("  Comment length: 0x%04x\n", end_of_central_directory_record.end_of_central_directory->comment_length);
    } else {
        printf("Error parsing end of central directory record\n");
    }

    free(data);

    return 0;
}

File: results/0.25/ZIP/zip-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::{
    bytes::complete::{take},
    combinator::{map_res},
    number::complete::{be_u16, be_u32},
    IResult,
};
use std::{
    env,
    fs::File,
    io::{BufReader, Read},
};

#[derive(Debug)]
enum CompressionMethod {
    Stored,
    Shrunk,
    Imploded,
    Tokenized,
    Deflated,
    ImplodedDeflated,
    BZip2,
    Lzma,
    Terse,
    Lz77,
    WavPack,
    PpmD,
    Aes128,
    Aes192,
    Aes256,
}

#[derive(Debug)]
struct LocalHeader {
    local_header_signature: u32,
    version_needed_to_extract: u16,
    general_purpose_bit_flag: u16,
    compression_method: CompressionMethod,
    last_mod_time: u16,
    last_mod_date: u16,
    crc_32: u32,
    compressed_size: u32,
    uncompressed_size: u32,
    filename_length: u16,
    extra_field_length: u16,
    filename: Vec<u8>,
    extra_field: Vec<u8>,
}

#[derive(Debug)]
struct CentralDirectoryHeader {
    central_header_signature: u32,
    version_made_by: u16,
    version_needed_to_extract: u16,
    general_purpose_bit_flag: u16,
    compression_method: CompressionMethod,
    last_mod_time: u16,
    last_mod_date: u16,
    crc_32: u32,
    compressed_size: u32,
    uncompressed_size: u32,
    filename_length: u16,
    extra_field_length: u16,
    file_comment_length: u16,
    disk_number_start: u16,
    int_attributes: u16,
    ext_attributes: u32,
    local_header_offset: u32,
    filename: Vec<u8>,
    extra_field: Vec<u8>,
    file_comment: Vec<u8>,
}

#[derive(Debug)]
struct EndOfCentralDirectoryRecord {
    end_of_central_dir_signature: u32,
    number_of_this_disk: u16,
    number_of_the_disk_with_the_start_of_the_central_directory: u16,
    total_number_of_entries_in_the_central_directory_on_this_disk: u16,
    total_number_of_entries_in_the_central_directory: u16,
    size_of_the_central_directory: u32,
    offset_of_start_of_central_directory_with_respect_to_the_starting_disk_number: u32,
    zipfile_comment_length: u16,
    zipfile_comment: Vec<u8>,
}

fn parse_local_header(input: &[u8]) -> IResult<&[u8], LocalHeader> {
    let (input, local_header_signature) = map_res(take(4usize), |x: &[u8]| {
        Ok(u32::from_be_bytes(x.try_into().unwrap()))
    })(input)?;
    assert_eq!(local_header_signature, 0x04034b50);
    let (input, version_needed_to_extract) = be_u16(input)?;
    let (input, general_purpose_bit_flag) = be_u16(input)?;
    let (input, compression_method) = map_res(be_u16, |x: u16| match x {
        0 => Ok(CompressionMethod::Stored),
        1 => Ok(CompressionMethod::Shrunk),
        2 => Ok(CompressionMethod::Imploded),
        3 => Ok(CompressionMethod::Tokenized),
        4 => Ok(CompressionMethod::Deflated),
        5 => Ok(CompressionMethod::ImplodedDeflated),
        6 => Ok(CompressionMethod::BZip2),
        7 => Ok(CompressionMethod::Lzma),
        8 => Ok(CompressionMethod::Terse),
        9 => Ok(CompressionMethod::Lz77),
        10 => Ok(CompressionMethod::WavPack),
        11 => Ok(CompressionMethod::PpmD),
        12 => Ok(CompressionMethod::Aes128),
        13 => Ok(CompressionMethod::Aes192),
        14 => Ok(CompressionMethod::Aes256),
        _ => Err(nom::Err::Error(nom::error::ErrorKind::Fail)),
    })(input)?;
    let (input, last_mod_time) = be_u16(input)?;
    let (input, last_mod_date) = be_u16(input)?;
    let (input, crc_32) = be_u32(input)?;
    let (input, compressed_size) = be_u32(input)?;
    let (input, uncompressed_size) = be_u32(input)?;
    let (input, filename_length) = be_u16(input)?;
    let (input, extra_field_length) = be_u16(input)?;
    let (input, filename) = take(filename_length as usize)(input)?;
    let (input, extra_field) = take(extra_field_length as usize)(input)?;
    Ok((
        input,
        LocalHeader {
            local_header_signature,
            version_needed_to_extract,
            general_purpose_bit_flag,
            compression_method,
            last_mod_time,
            last_mod_date,
            crc_32,
            compressed_size,
            uncompressed_size,
            filename_length,
            extra_field_length,
            filename: filename.to_vec(),
            extra_field: extra_field.to_vec(),
        },
    ))
}

fn parse_central_directory_header(input: &[u8]) -> IResult<&[u8], CentralDirectoryHeader> {
    let (input, central_header_signature) = map_res(take(4usize), |x: &[u8]| {
        Ok(u32::from_be_bytes(x.try_into().unwrap()))
    })(input)?;
    assert_eq!(central_header_signature, 0x02014b50);
    let (input, version_made_by) = be_u16(input)?;
    let (input, version_needed_to_extract) = be_u16(input)?;
    let (input, general_purpose_bit_flag) = be_u16(input)?;
    let (input, compression_method) = map_res(be_u16, |x: u16| match x {
        0 => Ok(CompressionMethod::Stored),
        1 => Ok(CompressionMethod::Shrunk),
        2 => Ok(CompressionMethod::Imploded),
        3 => Ok(CompressionMethod::Tokenized),
        4 => Ok(CompressionMethod::Deflated),
        5 => Ok(CompressionMethod::ImplodedDeflated),
        6 => Ok(CompressionMethod::BZip2),
        7 => Ok(CompressionMethod::Lzma),
        8 => Ok(CompressionMethod::Terse),
        9 => Ok(CompressionMethod::Lz77),
        10 => Ok(CompressionMethod::WavPack),
        11 => Ok(CompressionMethod::PpmD),
        12 => Ok(CompressionMethod::Aes128),
        13 => Ok(CompressionMethod::Aes192),
        14 => Ok(CompressionMethod::Aes256),
        _ => Err(nom::Err::Error(nom::error::ErrorKind::Fail)),
    })(input)?;
    let (input, last_mod_time) = be_u16(input)?;
    let (input, last_mod_date) = be_u16(input)?;
    let (input, crc_32) = be_u32(input)?;
    let (input, compressed_size) = be_u32(input)?;
    let (input, uncompressed_size) = be_u32(input)?;
    let (input, filename_length) = be_u16(input)?;
    let (input, extra_field_length) = be_u16(input)?;
    let (input, file_comment_length) = be_u16(input)?;
    let (input, disk_number_start) = be_u16(input)?;
    let (input, int_attributes) = be_u16(input)?;
    let (input, ext_attributes) = be_u32(input)?;
    let (input, local_header_offset) = be_u32(input)?;
    let (input, filename) = take(filename_length as usize)(input)?;
    let (input, extra_field) = take(extra_field_length as usize)(input)?;
    let (input, file_comment) = take(file_comment_length as usize)(input)?;
    Ok((
        input,
        CentralDirectoryHeader {
            central_header_signature,
            version_made_by,
            version_needed_to_extract,
            general_purpose_bit_flag,
            compression_method,
            last_mod_time,
            last_mod_date,
            crc_32,
            compressed_size,
            uncompressed_size,
            filename_length,
            extra_field_length,
            file_comment_length,
            disk_number_start,
            int_attributes,
            ext_attributes,
            local_header_offset,
            filename: filename.to_vec(),
            extra_field: extra_field.to_vec(),
            file_comment: file_comment.to_vec(),
        },
    ))
}

fn parse_end_of_central_directory_record(input: &[u8]) -> IResult<&[u8], EndOfCentralDirectoryRecord> {
    let (input, end_of_central_dir_signature) = map_res(take(4usize), |x: &[u8]| {
        Ok(u32::from_be_bytes(x.try_into().unwrap()))
    })(input)?;
    assert_eq!(end_of_central_dir_signature, 0x06054b50);
    let (input, number_of_this_disk) = be_u16(input)?;
    let (input, number_of_the_disk_with_the_start_of_the_central_directory) = be_u16(input)?;
    let (input, total_number_of_entries_in_the_central_directory_on_this_disk) = be_u16(input)?;
    let (input, total_number_of_entries_in_the_central_directory) = be_u16(input)?;
    let (input, size_of_the_central_directory) = be_u32(input)?;
    let (input, offset_of_start_of_central_directory_with_respect_to_the_starting_disk_number) = be_u32(input)?;
    let (input, zipfile_comment_length) = be_u16(input)?;
    let (input, zipfile_comment) = take(zipfile_comment_length as usize)(input)?;
    Ok((
        input,
        EndOfCentralDirectoryRecord {
            end_of_central_dir_signature,
            number_of_this_disk,
            number_of_the_disk_with_the_start_of_the_central_directory,
            total_number_of_entries_in_the_central_directory_on_this_disk,
            total_number_of_entries_in_the_central_directory,
            size_of_the_central_directory,
            offset_of_start_of_central_directory_with_respect_to_the_starting_disk_number,
            zipfile_comment_length,
            zipfile_comment: zipfile_comment.to_vec(),
        },
    ))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Usage: {} <input_file>", args[0]);
    }
    let file = File::open(&args[1]).unwrap();
    let mut reader = BufReader::new(file);
    let mut input = Vec::new();
    reader.read_to_end(&mut input).unwrap();
    let (input, local_header) = parse_local_header(&input).unwrap();
    println!("Local Header: {:?}", local_header);
    let (input, central_directory_header) = parse_central_directory_header(input).unwrap();
    println!("Central Directory Header: {:?}", central_directory_header);
    let (_input, end_of_central_directory_record) = parse_end_of_central_directory_record(input).unwrap();
    println!("End of Central Directory Record: {:?}", end_of_central_directory_record);
}